src/redisearch_rs/buffer/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+///
+/// * `buf` must point to a valid, properly initialized `Buffer`.
+/// * `pos` must not exceed the length of the Buffer.
+/// * The `BufferReader` does not own the `Buffer`, only borrows it.
+///
+/// It's best to use this through the `std::io::Read` implementation.
+#[repr(C)]
+pub struct BufferReader {
+    /// A pointer to a [`Buffer`]. If the `BufferReader` is coming from the C side, we must not
+    /// free or move it.
+    pub buf: NonNull<Buffer>,
+    /// The read position in the `buf` pointer.
+    pub pos: usize,
+}
+
+impl std::io::Read for BufferReader {
+    fn read(&mut self, dest_buf: &mut [u8]) -> std::io::Result<usize> {
+        // Safety: We assume `buf` is a valid pointer to a properly initialized Buffer.
+        let buffer = unsafe { self.buf.as_mut() };
+
+        // Safety: `self.pos` was just checked to be within the limits of the buffer's
+        // initialized range, so this pointer arithmetic is safe.
+        let src = unsafe { buffer.data.add(self.pos) };
+
+        // Safety: We just verified that `src` points to valid memory within the buffer.
+        let src = unsafe { src.as_ref() };
+        let dest = dest_buf.as_mut_ptr();
+        let len = cmp::min(dest_buf.len(), buffer.len - self.pos);
+
+        // Safety:
+        // * `src` points to initialized memory in the buffer.
+        // * `dest` is a valid pointer to the destination slice.
+        // * We've verified there are at least `dest_buf.len()` bytes available in the source.
+        // * The memory regions don't overlap (copy_nonoverlapping requirement).
+        unsafe { copy_nonoverlapping(src, dest, len) };
+
+        // Update position after successful read
+        self.pos += len;
+
+        Ok(len)
+    }
+}
+
+impl Buffer {
+    /// Creates a new `Buffer` with the given pointer, length, and capacity.
+    ///
+    /// # Safety
+    ///
+    /// * `data` must be a valid pointer to a memory region of at least `capacity` bytes.
+    /// * `len` must be less than or equal to `capacity`.
+    /// * The memory region must remain valid for the lifetime of the Buffer.
+    /// * The first `len` bytes of the memory region must be initialized.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `len` is greater than `capacity`.
+    pub unsafe fn new(data: NonNull<u8>, len: usize, capacity: usize) -> Self {
+        debug_assert!(len <= capacity, "len must not exceed capacity");
+        Self {
+            data,
+            len,
+            capacity,
+        }
+    }
+
+    /// Returns the initialized portion of the buffer as a slice.
+    pub fn as_slice(&self) -> &[u8] {
+        // Safety: We assume `self.data` is a valid pointer as per `Buffer`'s invariants.
+        let data = unsafe { self.data.as_ref() };
+
+        // Safety: We assume `self.len` is a valid length within the allocated memory.
+        // Creates a slice referencing the initialized portion of the buffer.
+        unsafe { slice::from_raw_parts(data, self.len) }
```

Comment:
```suggestion
        unsafe { slice::from_raw_parts(self.data.as_ptr(), self.len) }
```
Same as below, you don't want to create an intermediate reference.